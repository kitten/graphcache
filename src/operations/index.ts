export { query } from './query';
export { write, writeFragment } from './write';
export { gc } from './gc';
