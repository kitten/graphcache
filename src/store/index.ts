export {
  makeDict,
  initDataState,
  clearDataState,
  clearOptimistic,
  getCurrentDependencies,
} from './data';

export * from './keys';
export * from './store';
