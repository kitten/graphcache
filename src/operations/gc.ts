import { joinKeys } from '../helpers';
import { Store } from '../store';
import { Link } from '../types';

interface Context {
  store: Store;
  visitedLinks: Set<string>;
  visitedRecords: Set<string>;
}

export const gc = (store: Store) => {
  const visitedLinks = new Set<string>();
  const visitedRecords = new Set<string>();
  const ctx = { store, visitedLinks, visitedRecords };

  walkEntity(ctx, 'Query');

  store.records.forEach((_entity, key) => {
    if (!visitedRecords.has(key)) {
      store.remove(key);
    }
  });

  store.links.forEach((_link, key) => {
    if (!visitedLinks.has(key)) {
      store.removeLink(key);
    }
  });
};

const walkEntity = (ctx: Context, key: string) => {
  const { store, visitedRecords, visitedLinks } = ctx;
  const entity = store.find(key);

  if (entity !== null && !visitedRecords.has(key)) {
    visitedRecords.add(key);

    for (const fieldKey in entity) {
      const value = entity[fieldKey];
      if (value === undefined) {
        const childFieldKey = joinKeys(key, fieldKey);
        const link = store.readLink(childFieldKey);
        if (link !== undefined && !visitedLinks.has(childFieldKey)) {
          visitedLinks.add(childFieldKey);
          walkLink(ctx, link);
        }
      }
    }
  }
};

const walkLink = (ctx: Context, link: Link) => {
  if (Array.isArray(link)) {
    link.forEach(childLink => {
      walkLink(ctx, childLink);
    });
  } else if (link !== null) {
    walkEntity(ctx, link);
  }
};
