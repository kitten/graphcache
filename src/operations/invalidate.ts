import { FieldNode } from 'graphql';

import {
  getMainOperation,
  normalizeVariables,
  getFragments,
  getSelectionSet,
  getName,
  getFieldArguments,
} from '../ast';

import {
  EntityField,
  OperationRequest,
  Variables,
  Fragments,
  SelectionSet,
} from '../types';

import * as InMemoryData from '../store/data';
import { Store, keyOfField } from '../store';
import { SchemaPredicates } from '../ast';
import { SelectionIterator } from './shared';

interface Context {
  store: Store;
  variables: Variables;
  fragments: Fragments;
  schemaPredicates?: SchemaPredicates;
}

export const invalidate = (store: Store, request: OperationRequest) => {
  const operation = getMainOperation(request.query);

  const ctx: Context = {
    variables: normalizeVariables(operation, request.variables),
    fragments: getFragments(request.query),
    store,
    schemaPredicates: store.schemaPredicates,
  };

  invalidateSelection(
    ctx,
    ctx.store.getRootKey('query'),
    getSelectionSet(operation)
  );
};

export const invalidateSelection = (
  ctx: Context,
  entityKey: string,
  select: SelectionSet
) => {
  const isQuery = entityKey === 'Query';

  let typename: EntityField;
  if (!isQuery) {
    typename = InMemoryData.readRecord(entityKey, '__typename');
    if (typeof typename !== 'string') {
      return;
    } else {
      InMemoryData.writeRecord(entityKey, '__typename', undefined);
    }
  } else {
    typename = entityKey;
  }

  const iter = new SelectionIterator(typename, entityKey, select, ctx);

  let node: FieldNode | void;
  while ((node = iter.next()) !== undefined) {
    const fieldName = getName(node);
    const fieldKey = keyOfField(
      fieldName,
      getFieldArguments(node, ctx.variables)
    );

    if (
      process.env.NODE_ENV !== 'production' &&
      ctx.schemaPredicates &&
      typename
    ) {
      ctx.schemaPredicates.isFieldAvailableOnType(typename, fieldName);
    }

    if (node.selectionSet === undefined) {
      InMemoryData.writeRecord(entityKey, fieldKey, undefined);
    } else {
      const fieldSelect = getSelectionSet(node);
      const link = InMemoryData.readLink(entityKey, fieldKey);

      InMemoryData.writeLink(entityKey, fieldKey, undefined);
      InMemoryData.writeRecord(entityKey, fieldKey, undefined);

      if (Array.isArray(link)) {
        for (let i = 0, l = link.length; i < l; i++) {
          const childLink = link[i];
          if (childLink !== null) {
            invalidateSelection(ctx, childLink, fieldSelect);
          }
        }
      } else if (link) {
        invalidateSelection(ctx, link, fieldSelect);
      }
    }
  }
};
