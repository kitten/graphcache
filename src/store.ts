import invariant from 'invariant';
import { DocumentNode } from 'graphql';
import * as Pessimism from 'pessimism';

import {
  EntityField,
  Link,
  ResolverConfig,
  DataField,
  Variables,
  Data,
  UpdatesConfig,
  OptimisticMutationConfig,
  KeyingConfig,
} from './types';

import { keyOfEntity, joinKeys, keyOfField } from './helpers';
import { startQuery } from './operations/query';
import { writeFragment, startWrite } from './operations/write';
import { invalidate } from './operations/invalidate';

interface Ref<T> {
  current: null | T;
}

const currentDependencies: Ref<Set<string>> = { current: null };
const currentOptimisticKey: Ref<number> = { current: null };

// Resolve a ref value or throw when we're outside of a store run
const refValue = <T>(ref: Ref<T>): T => {
  invariant(
    ref.current !== null,
    'Invalid Cache call: The cache may only be accessed or mutated during' +
      'operations like write or query, or as part of its resolvers, updaters, ' +
      'or optimistic configs.'
  );

  return ref.current as T;
};

// Initialise a store run by resetting its internal state
export const initStoreState = (optimisticKey: null | number) => {
  currentDependencies.current = new Set();
  currentOptimisticKey.current = optimisticKey;
};

// Finalise a store run by clearing its internal state
export const clearStoreState = () => {
  currentDependencies.current = null;
  currentOptimisticKey.current = null;
};

export const getCurrentDependencies = () => refValue(currentDependencies);

// Add a dependency to the internal store state
export const addDependency = (dependency: string) => {
  (currentDependencies.current as Set<string>).add(dependency);
};

const mapSet = <T>(map: Pessimism.Map<T>, key: string, value: T) => {
  const optimisticKey = currentOptimisticKey.current || 0;
  return Pessimism.setOptimistic(map, key, value, optimisticKey);
};

// Used to remove a value from a Map optimistially (possible by setting it to undefined)
const mapRemove = <T>(map: Pessimism.Map<T>, key: string) => {
  const optimisticKey = currentOptimisticKey.current || 0;
  return optimisticKey
    ? Pessimism.setOptimistic(map, key, undefined, optimisticKey)
    : Pessimism.remove(map, key);
};

export class Store {
  records: Pessimism.Map<EntityField>;
  links: Pessimism.Map<Link>;

  resolvers: ResolverConfig;
  updates: UpdatesConfig;
  optimisticMutations: OptimisticMutationConfig;
  keys: KeyingConfig;

  constructor(
    resolvers?: ResolverConfig,
    updates?: Partial<UpdatesConfig>,
    optimisticMutations?: OptimisticMutationConfig,
    keys?: KeyingConfig
  ) {
    this.records = Pessimism.asMutable(Pessimism.make());
    this.links = Pessimism.asMutable(Pessimism.make());
    this.resolvers = resolvers || {};
    this.updates = {
      Mutation: (updates && updates.Mutation) || {},
      Subscription: (updates && updates.Subscription) || {},
    } as UpdatesConfig;
    this.optimisticMutations = optimisticMutations || {};
    this.keys = keys || {};
  }

  keyOfEntity(data: Data) {
    const { __typename: typename } = data;
    if (typename !== undefined && this.keys[typename] !== undefined) {
      return this.keys[typename](data);
    } else {
      return keyOfEntity(data);
    }
  }

  clearOptimistic(optimisticKey: number) {
    this.records = Pessimism.clearOptimistic(this.records, optimisticKey);
    this.links = Pessimism.clearOptimistic(this.links, optimisticKey);
  }

  getRecord(fieldKey: string): EntityField {
    return Pessimism.get(this.records, fieldKey);
  }

  removeRecord(fieldKey: string) {
    return (this.records = mapRemove(this.records, fieldKey));
  }

  writeRecord(field: EntityField, fieldKey: string) {
    return (this.records = mapSet(this.records, fieldKey, field));
  }

  getField(
    entityKey: string,
    fieldName: string,
    args?: Variables
  ): EntityField {
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, args));
    return this.getRecord(fieldKey);
  }

  writeField(
    field: EntityField,
    entityKey: string,
    fieldName: string,
    args?: Variables
  ) {
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, args));
    return (this.records = mapSet(this.records, fieldKey, field));
  }

  getLink(key: string): undefined | Link {
    return Pessimism.get(this.links, key);
  }

  removeLink(key: string) {
    return (this.links = mapRemove(this.links, key));
  }

  writeLink(link: Link, key: string) {
    return (this.links = mapSet(this.links, key, link));
  }

  resolveValueOrLink(fieldKey: string): DataField {
    const fieldValue = this.getRecord(fieldKey);
    // Undefined implies a link OR incomplete data.
    // A value will imply that we are just fetching a field like date.
    if (fieldValue !== undefined) return fieldValue;

    // This can be an array OR a string OR undefined again
    const link = this.getLink(fieldKey);
    return link ? link : null;
  }

  resolve(entity: Data | string, field: string, args?: Variables): DataField {
    if (typeof entity === 'string') {
      addDependency(entity);
      return this.resolveValueOrLink(joinKeys(entity, keyOfField(field, args)));
    } else {
      // This gives us __typename:key
      const entityKey = this.keyOfEntity(entity);
      if (entityKey === null) return null;
      addDependency(entityKey);
      return this.resolveValueOrLink(
        joinKeys(entityKey, keyOfField(field, args))
      );
    }
  }

  invalidateQuery(dataQuery: DocumentNode, variables: Variables) {
    invalidate(this, { query: dataQuery, variables });
  }

  hasField(key: string): boolean {
    return this.getRecord(key) !== undefined || this.getLink(key) !== undefined;
  }

  updateQuery(
    dataQuery: DocumentNode,
    updater: (data: Data | null) => null | Data
  ): void {
    const { data, completeness } = startQuery(this, { query: dataQuery });
    const input = completeness === 'EMPTY' ? null : data;
    const output = updater(input);
    if (output !== null) {
      startWrite(this, { query: dataQuery }, output);
    }
  }

  writeFragment(dataFragment: DocumentNode, data: Data): void {
    writeFragment(this, dataFragment, data);
  }
}
