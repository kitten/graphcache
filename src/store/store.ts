import { DocumentNode } from 'graphql';
import { createRequest } from 'urql/core';

import {
  Cache,
  FieldInfo,
  ResolverConfig,
  DataField,
  Variables,
  Data,
  QueryInput,
  UpdatesConfig,
  OptimisticMutationConfig,
  KeyingConfig,
} from '../types';

import { read, readFragment } from '../operations/query';
import { writeFragment, startWrite } from '../operations/write';
import { invalidate } from '../operations/invalidate';
import { SchemaPredicates } from '../ast';
import { keyOfField } from './keys';
import * as InMemoryData from './data';

type RootField = 'query' | 'mutation' | 'subscription';

export class Store implements Cache {
  data: InMemoryData.InMemoryData;

  resolvers: ResolverConfig;
  updates: UpdatesConfig;
  optimisticMutations: OptimisticMutationConfig;
  keys: KeyingConfig;
  schemaPredicates?: SchemaPredicates;

  rootFields: { query: string; mutation: string; subscription: string };
  rootNames: { [name: string]: RootField };

  constructor(
    schemaPredicates?: SchemaPredicates,
    resolvers?: ResolverConfig,
    updates?: Partial<UpdatesConfig>,
    optimisticMutations?: OptimisticMutationConfig,
    keys?: KeyingConfig
  ) {
    this.resolvers = resolvers || {};
    this.optimisticMutations = optimisticMutations || {};
    this.keys = keys || {};
    this.schemaPredicates = schemaPredicates;

    this.updates = {
      Mutation: (updates && updates.Mutation) || {},
      Subscription: (updates && updates.Subscription) || {},
    } as UpdatesConfig;

    if (schemaPredicates) {
      const { schema } = schemaPredicates;
      const queryType = schema.getQueryType();
      const mutationType = schema.getMutationType();
      const subscriptionType = schema.getSubscriptionType();

      const queryName = queryType ? queryType.name : 'Query';
      const mutationName = mutationType ? mutationType.name : 'Mutation';
      const subscriptionName = subscriptionType
        ? subscriptionType.name
        : 'Subscription';

      this.rootFields = {
        query: queryName,
        mutation: mutationName,
        subscription: subscriptionName,
      };

      this.rootNames = {
        [queryName]: 'query',
        [mutationName]: 'mutation',
        [subscriptionName]: 'subscription',
      };
    } else {
      this.rootFields = {
        query: 'Query',
        mutation: 'Mutation',
        subscription: 'Subscription',
      };

      this.rootNames = {
        Query: 'query',
        Mutation: 'mutation',
        Subscription: 'subscription',
      };
    }

    this.data = InMemoryData.make(this.getRootKey('query'));
  }

  gcScheduled = false;
  gc = () => {
    InMemoryData.gc(this.data);
    this.gcScheduled = false;
  };

  keyOfField = keyOfField;

  getRootKey(name: RootField) {
    return this.rootFields[name];
  }

  keyOfEntity(data: Data) {
    const { __typename: typename, id, _id } = data;
    if (!typename) {
      return null;
    } else if (this.rootNames[typename] !== undefined) {
      return typename;
    }

    let key: string | null | void;
    if (this.keys[typename]) {
      key = this.keys[typename](data);
    } else if (id !== undefined && id !== null) {
      key = `${id}`;
    } else if (_id !== undefined && _id !== null) {
      key = `${_id}`;
    }

    return key ? `${typename}:${key}` : null;
  }

  resolveFieldByKey(entity: Data | string | null, fieldKey: string): DataField {
    const entityKey =
      entity !== null && typeof entity !== 'string'
        ? this.keyOfEntity(entity)
        : entity;
    if (entityKey === null) return null;
    const fieldValue = InMemoryData.readRecord(entityKey, fieldKey);
    if (fieldValue !== undefined) return fieldValue;
    const link = InMemoryData.readLink(entityKey, fieldKey);
    return link ? link : null;
  }

  resolve(
    entity: Data | string | null,
    field: string,
    args?: Variables
  ): DataField {
    return this.resolveFieldByKey(entity, keyOfField(field, args));
  }

  invalidateQuery(query: string | DocumentNode, variables?: Variables) {
    invalidate(this, createRequest(query, variables));
  }

  inspectFields(entity: Data | string | null): FieldInfo[] {
    const entityKey =
      entity !== null && typeof entity !== 'string'
        ? this.keyOfEntity(entity)
        : entity;
    return entityKey !== null ? InMemoryData.inspectFields(entityKey) : [];
  }

  updateQuery(
    input: QueryInput,
    updater: (data: Data | null) => Data | null
  ): void {
    const request = createRequest(input.query, input.variables);
    const output = updater(this.readQuery(request as QueryInput));
    if (output !== null) {
      startWrite(this, request, output);
    }
  }

  readQuery(input: QueryInput): Data | null {
    return read(this, createRequest(input.query, input.variables)).data;
  }

  readFragment(
    dataFragment: DocumentNode,
    entity: string | Data,
    variables?: Variables
  ): Data | null {
    return readFragment(this, dataFragment, entity, variables);
  }

  writeFragment(
    dataFragment: DocumentNode,
    data: Data,
    variables?: Variables
  ): void {
    writeFragment(this, dataFragment, data, variables);
  }
}
