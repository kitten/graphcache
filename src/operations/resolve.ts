import { Store } from '../store';
import {
  Data,
  SelectionSet,
  OperationRequest,
  Variables,
  Fragments,
  DataField,
} from '../types';
import {
  getSelectionSet,
  getMainOperation,
  normalizeVariables,
  getFragments,
  getName,
  getFieldArguments,
  getFieldAlias,
} from '../ast';
import { SelectionIterator } from './shared';
import { joinKeys, keyOfField, isDataOrKey } from '../helpers';
import warning from 'warning';

interface Context {
  store: Store;
  variables: Variables;
  fragments: Fragments;
}

export const resolveData = (
  store: Store,
  request: OperationRequest,
  data: Data
) => {
  const operation = getMainOperation(request.query);
  const ctx = {
    store,
    variables: normalizeVariables(operation, request.variables),
    fragments: getFragments(request.query),
  };
  resolveDataSelection(ctx, 'Query', getSelectionSet(operation), data);
  return data;
};

const resolveResolverResult = (
  ctx: Context,
  result: DataField,
  key: string,
  select: SelectionSet,
  prevData: void | Data | Data[]
) => {
  // When we are dealing with a list we have to call this method again.
  if (Array.isArray(result)) {
    // @ts-ignore: Link cannot be expressed as a recursive type
    return result.map((childResult, index) => {
      const data = prevData !== undefined ? prevData[index] : undefined;
      const indexKey = joinKeys(key, `${index}`);
      return resolveResolverResult(ctx, childResult, indexKey, select, data);
    });
  } else if (result === null) {
    return null;
  } else if (isDataOrKey(result)) {
    // We don't need to read the entity after exiting a resolver
    // we can just go on and read the selection further.
    const data = prevData === undefined ? Object.create(null) : prevData;
    const childKey =
      (typeof result === 'string' ? result : ctx.store.keyOfEntity(result)) ||
      key;
    const selectionResult = resolveDataSelection(ctx, childKey, select, data);

    if (selectionResult !== null && typeof result === 'object') {
      for (key in result) {
        if (key !== '__typename' && result.hasOwnProperty(key)) {
          selectionResult[key] = result[key];
        }
      }
    }

    return selectionResult;
  }

  warning(
    false,
    'Invalid resolver value: The resolver at `%s` returned a scalar (number, boolean, etc)' +
      ', but the GraphQL query expects a selection set for this field.\n' +
      'If necessary, use Cache.resolve() to resolve a link or entity from the cache.',
    key
  );

  return null;
};

const resolveDataSelection = (
  ctx: Context,
  entityKey: string,
  select: SelectionSet,
  data: Data
) => {
  const { store, variables } = ctx;
  const typename =
    entityKey === 'Query'
      ? entityKey
      : (store.getField(entityKey, '__typename') as string);
  const iter = new SelectionIterator(typename, entityKey, select, ctx);

  let node;
  while ((node = iter.next()) !== undefined) {
    // Derive the needed data from our node.
    const fieldName = getName(node);
    const fieldArgs = getFieldArguments(node, variables);
    const fieldAlias = getFieldAlias(node);
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, fieldArgs));

    const resolvers = store.resolvers[typename];
    if (resolvers !== undefined && resolvers.hasOwnProperty(fieldName)) {
      // We have a resolver for this field.
      const resolverValue = resolvers[fieldName](
        data,
        fieldArgs || {},
        store,
        ctx
      );

      if (node.selectionSet === undefined) {
        data[fieldAlias] = resolverValue !== undefined ? resolverValue : null;
      } else {
        const fieldSelect = getSelectionSet(node);

        data[fieldAlias] = resolveResolverResult(
          ctx,
          resolverValue,
          fieldKey,
          fieldSelect,
          data[fieldAlias] as Data | Data[]
        );
      }
    }
  }
};
