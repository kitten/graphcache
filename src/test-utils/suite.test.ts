import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';
import { query, write } from '../operations';
import { Store } from '../store';

interface TestCase {
  query: DocumentNode;
  variables?: any;
  data: any;
}

const expectCacheIntegrity = (testcase: TestCase) => {
  const store = new Store();
  const request = { query: testcase.query, variables: testcase.variables };
  const writeRes = write(store, request, testcase.data);
  const queryRes = query(store, request);
  expect(queryRes.data).toEqual(testcase.data);
  expect(queryRes.completeness).toBe('FULL');
  expect(queryRes.dependencies).toEqual(writeRes.dependencies);
};

it('int on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        int
      }
    `,
    data: { __typename: 'Query', int: 42 },
  });
});

it('aliased field on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        anotherName: int
      }
    `,
    data: { __typename: 'Query', anotherName: 42 },
  });
});

it('json on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        json
      }
    `,
    // The `__typename` field should not mislead the cache
    data: {
      __typename: 'Query',
      json: { __typename: 'Misleading', test: true },
    },
  });
});

it('nullable field on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        missing
      }
    `,
    data: { __typename: 'Query', missing: null },
  });
});

it('int field with arguments on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        int(test: true)
      }
    `,
    data: { __typename: 'Query', int: 42 },
  });
});

it('non-keyable entity on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item {
          __typename
          name
        }
      }
    `,
    // This entity has no `id` or `_id` field
    data: { __typename: 'Query', item: { __typename: 'Item', name: 'Test' } },
  });
});

it('invalid entity on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item {
          __typename
          id
          name
        }
      }
    `,
    // This entity comes back with an invalid typename (for some reason or another)
    data: {
      __typename: 'Query',
      item: { __typename: null, id: '123', name: 'Test' },
    },
  });
});

it('non-IDable entity on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item {
          __typename
          name
        }
      }
    `,
    // This entity has a `__typename` but no ID fields
    data: { __typename: 'Query', item: { __typename: 'Item', name: 'Test' } },
  });
});

it('entity on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item {
          __typename
          id
          name
        }
      }
    `,
    data: {
      __typename: 'Query',
      item: { __typename: 'Item', id: '1', name: 'Test' },
    },
  });
});

it('entity on aliased field on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        anotherName: item {
          __typename
          id
          name
        }
      }
    `,
    data: {
      __typename: 'Query',
      anotherName: { __typename: 'Item', id: '1', name: 'Test' },
    },
  });
});

it('entity with arguments on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item(test: true) {
          __typename
          id
          name
        }
      }
    `,
    data: {
      __typename: 'Query',
      item: { __typename: 'Item', id: '1', name: 'Test' },
    },
  });
});

it('entity with Int-like ID on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item {
          __typename
          id
          name
        }
      }
    `,
    // This is the same as above, but with a number on `id`
    data: {
      __typename: 'Query',
      item: { __typename: 'Item', id: 1, name: 'Test' },
    },
  });
});

it('entity list on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        items {
          __typename
          id
        }
      }
    `,
    data: {
      __typename: 'Query',
      items: [{ __typename: 'Item', id: 1 }, { __typename: 'Item', id: 2 }],
    },
  });
});

it('nested entity list on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        items {
          __typename
          id
        }
      }
    `,
    data: {
      // This is the same as above, but with a nested array and added null values
      __typename: 'Query',
      items: [
        { __typename: 'Item', id: 1 },
        [{ __typename: 'Item', id: 2 }, null],
        null,
      ],
    },
  });
});

it('entity list on query and inline fragment', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        items {
          __typename
          id
        }
        ... on Query {
          items {
            test
          }
        }
      }
    `,
    data: {
      __typename: 'Query',
      items: [{ __typename: 'Item', id: 1, test: true }, null],
    },
  });
});

it('entity list on query and spread fragment', () => {
  expectCacheIntegrity({
    query: gql`
      query Test {
        __typename
        items {
          __typename
          id
        }
        ...TestFragment
      }

      fragment TestFragment on Query {
        items {
          test
        }
      }
    `,
    data: {
      __typename: 'Query',
      items: [{ __typename: 'Item', id: 1, test: true }, null],
    },
  });
});

it('embedded object on entity', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item {
          __typename
          id
          author {
            __typename
            name
          }
        }
      }
    `,
    data: {
      __typename: 'Query',
      item: {
        __typename: 'Item',
        id: 1,
        author: {
          __typename: 'Author',
          name: 'Stanley',
        },
      },
    },
  });
});

it('embedded object on entity', () => {
  expectCacheIntegrity({
    query: gql`
      {
        __typename
        item {
          __typename
          id
          author {
            __typename
            id
            name
          }
        }
      }
    `,
    data: {
      __typename: 'Query',
      item: {
        __typename: 'Item',
        id: 1,
        author: {
          __typename: 'Author',
          id: 1,
          name: 'Stanley',
        },
      },
    },
  });
});
