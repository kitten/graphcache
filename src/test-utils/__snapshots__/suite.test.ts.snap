// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aliased field on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "int": 42,
    },
  },
}
`;

exports[`aliased field on query 2`] = `
Set {
  "Query.int",
}
`;

exports[`embedded object on entity 1`] = `
Object {
  "links": Object {
    "Item:1.author": "Item:1.author",
    "Query.item": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "author": undefined,
      "id": 1,
    },
    "Item:1.author": Object {
      "__typename": "Author",
      "name": "Stanley",
    },
    "Query": Object {
      "__typename": "Query",
      "item": undefined,
    },
  },
}
`;

exports[`embedded object on entity 2`] = `
Set {
  "Query.item",
  "Item:1",
  "Item:1.author",
}
`;

exports[`embedded object on entity 3`] = `
Object {
  "links": Object {
    "Item:1.author": "Author:1",
    "Query.item": "Item:1",
  },
  "records": Object {
    "Author:1": Object {
      "__typename": "Author",
      "id": 1,
      "name": "Stanley",
    },
    "Item:1": Object {
      "__typename": "Item",
      "author": undefined,
      "id": 1,
    },
    "Query": Object {
      "__typename": "Query",
      "item": undefined,
    },
  },
}
`;

exports[`embedded object on entity 4`] = `
Set {
  "Query.item",
  "Item:1",
  "Author:1",
}
`;

exports[`entity list on query 1`] = `
Object {
  "links": Object {
    "Query.items": Array [
      "Item:1",
      "Item:2",
    ],
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
    },
    "Item:2": Object {
      "__typename": "Item",
      "id": 2,
    },
    "Query": Object {
      "__typename": "Query",
      "items": undefined,
    },
  },
}
`;

exports[`entity list on query 2`] = `
Set {
  "Query.items",
  "Item:1",
  "Item:2",
}
`;

exports[`entity list on query and inline fragment 1`] = `
Object {
  "links": Object {
    "Query.items": Array [
      "Item:1",
      null,
    ],
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
      "test": true,
    },
    "Query": Object {
      "__typename": "Query",
      "items": undefined,
    },
  },
}
`;

exports[`entity list on query and inline fragment 2`] = `
Set {
  "Query.items",
  "Item:1",
}
`;

exports[`entity list on query and spread fragment 1`] = `
Object {
  "links": Object {
    "Query.items": Array [
      "Item:1",
      null,
    ],
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
      "test": true,
    },
    "Query": Object {
      "__typename": "Query",
      "items": undefined,
    },
  },
}
`;

exports[`entity list on query and spread fragment 2`] = `
Set {
  "Query.items",
  "Item:1",
}
`;

exports[`entity on aliased field on query 1`] = `
Object {
  "links": Object {
    "Query.item": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": "1",
      "name": "Test",
    },
    "Query": Object {
      "__typename": "Query",
      "item": undefined,
    },
  },
}
`;

exports[`entity on aliased field on query 2`] = `
Set {
  "Query.item",
  "Item:1",
}
`;

exports[`entity on query 1`] = `
Object {
  "links": Object {
    "Query.item": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": "1",
      "name": "Test",
    },
    "Query": Object {
      "__typename": "Query",
      "item": undefined,
    },
  },
}
`;

exports[`entity on query 2`] = `
Set {
  "Query.item",
  "Item:1",
}
`;

exports[`entity with Int-like ID on query 1`] = `
Object {
  "links": Object {
    "Query.item": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
      "name": "Test",
    },
    "Query": Object {
      "__typename": "Query",
      "item": undefined,
    },
  },
}
`;

exports[`entity with Int-like ID on query 2`] = `
Set {
  "Query.item",
  "Item:1",
}
`;

exports[`entity with arguments on query 1`] = `
Object {
  "links": Object {
    "Query.item({\\"test\\":true})": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": "1",
      "name": "Test",
    },
    "Query": Object {
      "__typename": "Query",
      "item({\\"test\\":true})": undefined,
    },
  },
}
`;

exports[`entity with arguments on query 2`] = `
Set {
  "Query.item({\\"test\\":true})",
  "Item:1",
}
`;

exports[`int field with arguments on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "int({\\"test\\":true})": 42,
    },
  },
}
`;

exports[`int field with arguments on query 2`] = `
Set {
  "Query.int({\\"test\\":true})",
}
`;

exports[`int on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "int": 42,
    },
  },
}
`;

exports[`int on query 2`] = `
Set {
  "Query.int",
}
`;

exports[`invalid entity on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "item": Object {
        "__typename": null,
        "id": "123",
        "name": "Test",
      },
    },
  },
}
`;

exports[`invalid entity on query 2`] = `
Set {
  "Query.item",
}
`;

exports[`json on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "json": Object {
        "__typename": "Misleading",
        "test": true,
      },
    },
  },
}
`;

exports[`json on query 2`] = `
Set {
  "Query.json",
}
`;

exports[`nested entity list on query 1`] = `
Object {
  "links": Object {
    "Query.items": Array [
      "Item:1",
      Array [
        "Item:2",
        null,
      ],
      null,
    ],
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
    },
    "Item:2": Object {
      "__typename": "Item",
      "id": 2,
    },
    "Query": Object {
      "__typename": "Query",
      "items": undefined,
    },
  },
}
`;

exports[`nested entity list on query 2`] = `
Set {
  "Query.items",
  "Item:1",
  "Item:2",
}
`;

exports[`non-IDable entity on query 1`] = `
Object {
  "links": Object {
    "Query.item": "Query.item",
  },
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "item": undefined,
    },
    "Query.item": Object {
      "__typename": "Item",
      "name": "Test",
    },
  },
}
`;

exports[`non-IDable entity on query 2`] = `
Set {
  "Query.item",
}
`;

exports[`non-keyable entity on query 1`] = `
Object {
  "links": Object {
    "Query.item": "Query.item",
  },
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "item": undefined,
    },
    "Query.item": Object {
      "__typename": "Item",
      "name": "Test",
    },
  },
}
`;

exports[`non-keyable entity on query 2`] = `
Set {
  "Query.item",
}
`;

exports[`nullable field on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "missing": null,
    },
  },
}
`;

exports[`nullable field on query 2`] = `
Set {
  "Query.missing",
}
`;
